File: src/vca/mcp/manager.py
"""MCP 管理器 - 配置读取、连接 MCP servers、收集动态工具

利用 langchain-mcp-adapters 的 connection 机制：
每个 MCP 工具在调用时自动创建/销毁 session，无需维护长连接。

配置文件格式 (mcp.json):
{
  "servers": {
    "server_name": {
      "transport": "stdio",            // stdio | http | sse | websocket
      "command": "npx",                // stdio 必需
      "args": ["-y", "@modelcontextprotocol/server-fetch"],
      "env": {"KEY": "VALUE"}          // 可选
    },
    "http_server": {
      "transport": "http",             // http/sse 使用 url
      "url": "http://localhost:8000/mcp"
    }
  }
}

配置文件位置 (优先级从高到低):
1. 项目根:  <workspace>/.vca/mcp.json
2. 用户级:  ~/.vca/mcp.json
"""

from __future__ import annotations

import asyncio
import json
import os
from pathlib import Path
from typing import Any

from langchain_core.tools import StructuredTool
from langchain_mcp_adapters.tools import (
    convert_mcp_tool_to_langchain_tool,
    create_session,
)
from mcp.types import Tool as MCPTool


def _make_sync_compatible(tool: StructuredTool) -> StructuredTool:
    """
    给只有 coroutine 的 StructuredTool 补充同步 func。

    新版 langchain-core 的 StructuredTool 若只有 coroutine，
    同步调用会抛 "StructuredTool does not support sync invocation"。
    通过 asyncio.run 包装 coroutine 为同步函数解决。
    """
    original_coro = tool.coroutine

    async def _arun(**kwargs: Any) -> Any:
        return await original_coro(**kwargs)

    def _run(**kwargs: Any) -> Any:
        try:
            return asyncio.run(original_coro(**kwargs))
        except RuntimeError:
            # 已有 running loop (async 环境中), 走事件循环执行
            return asyncio.get_event_loop().run_until_complete(
                original_coro(**kwargs)
            )

    return StructuredTool.from_function(
        name=tool.name,
        description=tool.description,
        func=_run,
        coroutine=_arun,
        args_schema=tool.args_schema,
    )


class MCPManager:
    """管理所有 MCP server 连接与动态工具"""

    def __init__(self) -> None:
        self._tools: list[Any] = []
        self._server_status: dict[str, str] = {}
        self._connected = False

    # --------------------------------------------------------
    # 配置
    # --------------------------------------------------------

    def _load_config(self) -> dict[str, dict[str, Any]]:
        """读取 MCP 配置 (项目级优先, 合并用户级)"""
        servers: dict[str, dict[str, Any]] = {}

        candidates = [
            Path.home() / ".vca" / "mcp.json",
            Path.cwd() / ".vca" / "mcp.json",
        ]
        for path in candidates:
            if path.exists():
                try:
                    data = json.loads(path.read_text(encoding="utf-8"))
                    cfg = data.get("servers", {})
                    if isinstance(cfg, dict):
                        servers.update(cfg)
                except Exception:
                    continue

        # 兼容旧格式: 无 transport 字段时默认 stdio
        for name, cfg in servers.items():
            if "transport" not in cfg and "url" not in cfg:
                cfg["transport"] = "stdio"
            if cfg.get("transport") in ("http", "sse", "websocket") and "url" not in cfg:
                # 无 url 视为无效配置
                pass

        return servers

    # --------------------------------------------------------
    # 连接与工具收集
    # --------------------------------------------------------

    async def _list_server_tools(self, cfg: dict[str, Any]) -> list[MCPTool]:
        """连接 server 并列出其工具 (短生命周期)"""
        async with create_session(cfg) as session:
            await session.initialize()
            result = await session.list_tools()
            return result.tools

    def connect(self) -> dict[str, str]:
        """
        连接所有配置的 MCP servers 并收集工具。

        Returns:
            {server_name: "ok (N tools)" | "error: ..."}
        """
        servers = self._load_config()
        if not servers:
            self._connected = True
            return {}

        status: dict[str, str] = {}
        all_tools: list[Any] = []
        multi_server = len(servers) > 1

        for name, cfg in servers.items():
            try:
                loop = asyncio.new_event_loop()
                try:
                    asyncio.set_event_loop(loop)
                    tools_list = loop.run_until_complete(self._list_server_tools(cfg))
                finally:
                    loop.close()
                    asyncio.set_event_loop(None)

                # 将 MCP 工具转换为 LangChain 工具 (调用时自动建 session)
                converted = []
                for t in tools_list:
                    lc_tool = convert_mcp_tool_to_langchain_tool(
                        session=None,
                        tool=t,
                        connection=cfg,
                        server_name=name,
                        tool_name_prefix=multi_server,
                    )
                    # 补充同步支持 (同步图需要)
                    converted.append(_make_sync_compatible(lc_tool))
                all_tools.extend(converted)
                status[name] = f"ok ({len(converted)} tools)"
            except Exception as exc:
                status[name] = f"error: {type(exc).__name__}: {exc}"

        self._tools = all_tools
        self._server_status = status
        self._connected = True
        return status

    # --------------------------------------------------------
    # 查询
    # --------------------------------------------------------

    @property
    def tools(self) -> list[Any]:
        """已收集的所有 MCP 工具"""
        return self._tools

    @property
    def is_connected(self) -> bool:
        return self._connected

    def status_text(self) -> str:
        """生成状态文本 (供 /mcp 命令显示)"""
        if not self._server_status:
            return (
                "MCP servers 未配置。\n"
                "在 ~/.vca/mcp.json 或 <项目>/.vca/mcp.json 中配置后重启生效。"
            )
        lines = ["MCP server 状态:"]
        for name, st in self._server_status.items():
            lines.append(f"  - {name}: {st}")
        if self._tools:
            names = ", ".join(t.name for t in self._tools)
            lines.append(f"共加载 {len(self._tools)} 个工具: {names}")
        return "\n".join(lines)


# 全局单例
mcp_manager = MCPManager()

File: src/mcp/manager.ts
/**
 * MCP 管理器 - 配置读取、连接 MCP servers、收集动态工具
 *
 * 每个 MCP 工具在调用时自动创建/销毁 session (短生命周期, 参考 Python 版)。
 *
 * 配置文件格式 (mcp.json):
 * {
 *   "servers": {
 *     "server_name": {
 *       "transport": "stdio",                 // stdio | http
 *       "command": "npx",                     // stdio 必需
 *       "args": ["-y", "@modelcontextprotocol/server-fetch"],
 *       "env": {"KEY": "VALUE"}               // 可选
 *     },
 *     "http_server": {
 *       "transport": "http",                  // http 使用 url
 *       "url": "http://localhost:8000/mcp"
 *     }
 *   }
 * }
 *
 * 配置位置 (优先级从高到低):
 * 1. 项目根: <workspace>/.vca/mcp.json
 * 2. 用户级: ~/.vca/mcp.json
 */
import fs from "node:fs";
import path from "node:path";
import { DynamicStructuredTool, type StructuredToolInterface } from "@langchain/core/tools";
import { z } from "zod";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";

import { MCP_CONFIG_FILE } from "../config.js";
import { getWorkspace } from "../workspace_ctx.js";

export interface McpServerConfig {
  transport?: string;
  command?: string;
  args?: string[];
  env?: Record<string, string>;
  url?: string;
  [key: string]: unknown;
}

interface McpServer {
  name: string;
  config: McpServerConfig;
}

// ============================================================
// JSON Schema → Zod 转换 (MCP 工具 inputSchema)
// ============================================================

function jsonSchemaToZod(schema: Record<string, unknown>): z.ZodType {
  const type = (schema.type as string) ?? inferType(schema);

  switch (type) {
    case "string":
      return z.string();
    case "number":
      return z.number();
    case "integer":
      return z.number().int();
    case "boolean":
      return z.boolean();
    case "null":
      return z.null();
    case "array": {
      const items = schema.items as Record<string, unknown> | undefined;
      return z.array(items ? jsonSchemaToZod(items) : z.any());
    }
    case "object": {
      const properties = (schema.properties as Record<string, unknown>) ?? {};
      const required = (schema.required as string[]) ?? [];
      const shape: Record<string, z.ZodType> = {};
      for (const [key, value] of Object.entries(properties)) {
        const fieldSchema = jsonSchemaToZod(value as Record<string, unknown>);
        shape[key] = required.includes(key) ? fieldSchema : fieldSchema.optional();
      }
      return z.object(shape);
    }
    case "enum": {
      const enums = schema.enum as unknown[] | undefined;
      if (enums && enums.length > 0 && enums.every((e) => typeof e === "string")) {
        return z.enum(enums as [string, ...string[]]);
      }
      return z.any();
    }
    default:
      return z.any();
  }
}

function inferType(schema: Record<string, unknown>): string {
  if (schema.enum) return "enum";
  if (schema.properties) return "object";
  if (schema.items) return "array";
  return "string";
}

// ============================================================
// 会话创建与调用
// ============================================================

async function createClient(server: McpServer): Promise<{ client: Client; close: () => Promise<void> }> {
  const cfg = server.config;
  const client = new Client(
    { name: "vca", version: "0.2.0" },
    { capabilities: {} }
  );

  if (cfg.transport === "http" || cfg.transport === "sse" || cfg.transport === "streamable-http") {
    if (!cfg.url) throw new Error("http/sse transport 需要 url 配置");
    const transport = new StreamableHTTPClientTransport(new URL(cfg.url));
    await client.connect(transport);
    return {
      client,
      close: async () => {
        try {
          await client.close();
          await transport.close();
        } catch {
          /* ignore */
        }
      },
    };
  }

  // 默认 stdio
  const command = cfg.command;
  if (!command) throw new Error("stdio transport 需要 command 配置");
  const transport = new StdioClientTransport({
    command,
    args: cfg.args ?? [],
    env: cfg.env ?? {},
    cwd: getWorkspace(),
    stderr: "pipe",
  });
  await client.connect(transport);
  return {
    client,
    close: async () => {
      try {
        await client.close();
        await transport.close();
      } catch {
        /* ignore */
      }
    },
  };
}

async function callMcpTool(server: McpServer, toolName: string, args: Record<string, unknown>): Promise<string> {
  const { client, close } = await createClient(server);
  try {
    const result = await client.callTool({ name: toolName, arguments: args });
    return formatMcpResult(result);
  } finally {
    await close();
  }
}

function formatMcpResult(result: unknown): string {
  const r = result as {
    content?: Array<{ type?: string; text?: string }>;
    isError?: boolean;
    structuredContent?: unknown;
  };
  const parts: string[] = [];
  for (const block of r.content ?? []) {
    if (block.text) parts.push(block.text);
    else parts.push(JSON.stringify(block));
  }
  if (r.structuredContent !== undefined) {
    parts.push(JSON.stringify(r.structuredContent, null, 2));
  }
  const body = parts.join("\n").trim() || "(空结果)";
  return r.isError ? `[MCP ERROR] ${body}` : `[MCP OK] ${body}`;
}

// ============================================================
// MCPManager
// ============================================================

export class MCPManager {
  private _tools: StructuredToolInterface[] = [];
  private _serverStatus = new Map<string, string>();
  private _connected = false;

  /** 读取 MCP 配置 (用户级 + 项目级, 后者覆盖前者) */
  loadConfig(): McpServer[] {
    const servers = new Map<string, McpServerConfig>();
    const candidates = [
      MCP_CONFIG_FILE,
      path.join(getWorkspace(), ".vca", "mcp.json"),
    ];
    for (const file of candidates) {
      try {
        if (!fs.existsSync(file)) continue;
        const data = JSON.parse(fs.readFileSync(file, "utf-8")) as {
          servers?: Record<string, McpServerConfig>;
        };
        const cfg = data.servers ?? {};
        for (const [name, serverCfg] of Object.entries(cfg)) {
          if (serverCfg && typeof serverCfg === "object") {
            servers.set(name, serverCfg);
          }
        }
      } catch {
        continue;
      }
    }
    return [...servers.entries()].map(([name, config]) => ({ name, config }));
  }

  /** 连接所有配置的 servers 并收集工具 (失败不阻塞) */
  async connect(): Promise<Record<string, string>> {
    const servers = this.loadConfig();
    const status: Record<string, string> = {};
    const allTools: StructuredToolInterface[] = [];
    const multiServer = servers.length > 1;

    for (const server of servers) {
      try {
        const { client, close } = await createClient(server);
        let tools: Array<{ name: string; description?: string; inputSchema: Record<string, unknown> }> = [];
        try {
          const result = await client.listTools();
          tools = result.tools.map((t) => ({
            name: t.name,
            description: t.description ?? "",
            inputSchema: (t.inputSchema as Record<string, unknown>) ?? {},
          }));
        } finally {
          await close();
        }

        const converted = tools.map((t) => this.toLangChainTool(server, t, multiServer));
        allTools.push(...converted);
        status[server.name] = `ok (${converted.length} tools)`;
      } catch (e) {
        status[server.name] = `error: ${(e as Error).constructor.name}: ${(e as Error).message}`;
      }
    }

    this._tools = allTools;
    this._serverStatus = new Map(Object.entries(status));
    this._connected = true;
    return status;
  }

  private toLangChainTool(
    server: McpServer,
    mcpTool: { name: string; description?: string; inputSchema: Record<string, unknown> },
    multiServer: boolean
  ): StructuredToolInterface {
    const toolName = multiServer ? `${server.name}__${mcpTool.name}` : mcpTool.name;
    const schema = jsonSchemaToZod(mcpTool.inputSchema);

    return new DynamicStructuredTool({
      name: toolName,
      description: `[MCP:${server.name}] ${mcpTool.description ?? mcpTool.name}`,
      schema,
      func: async (args) => {
        return callMcpTool(server, mcpTool.name, args as Record<string, unknown>);
      },
    });
  }

  get tools(): StructuredToolInterface[] {
    return this._tools;
  }

  get isConnected(): boolean {
    return this._connected;
  }

  /** server 状态 Map (name → 状态文本) */
  serverStatus(): Map<string, string> {
    return new Map(this._serverStatus);
  }

  statusText(): string {
    if (this._serverStatus.size === 0) {
      return (
        "MCP servers 未配置。\n" +
        "在 ~/.vca/mcp.json 或 <项目>/.vca/mcp.json 中配置后重启生效。"
      );
    }
    const lines = ["MCP server 状态:"];
    for (const [name, st] of this._serverStatus) {
      lines.push(`  - ${name}: ${st}`);
    }
    if (this._tools.length > 0) {
      const names = this._tools.map((t) => t.name).join(", ");
      lines.push(`共加载 ${this._tools.length} 个工具: ${names}`);
    }
    return lines.join("\n");
  }
}

// 全局单例
export const mcpManager = new MCPManager();
